pkg/kubectl/cmd/util/factory.go

```diff
@@ -688,21 +696,49 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 	}
 }
 
-// GetFirstPod returns the first pod of an object from its namespace and selector and the number of matching pods
-func GetFirstPod(client *client.Client, namespace string, selector labels.Selector) (*api.Pod, int, error) {
-	var pods *api.PodList
-	for pods == nil || len(pods.Items) == 0 {
-		var err error
-		options := api.ListOptions{LabelSelector: selector}
-		if pods, err = client.Pods(namespace).List(options); err != nil {
-			return nil, 0, err
-		}
-		if len(pods.Items) == 0 {
-			time.Sleep(2 * time.Second)
+// GetFirstPod returns a pod matching the namespace and label selector
+// and the number of all pods that match the label selector.
+func GetFirstPod(client client.Interface, namespace string, selector labels.Selector, timeout time.Duration, sortBy func([]*api.Pod) sort.Interface) (*api.Pod, int, error) {
+	options := api.ListOptions{LabelSelector: selector}
+
+	podList, err := client.Pods(namespace).List(options)
+	if err != nil {
+		return nil, 0, err
+	}
+	pods := []*api.Pod{}
+	for i := range podList.Items {
+		pod := podList.Items[i]
+		pods = append(pods, &pod)
+	}
+	if len(pods) > 0 {
+		sort.Sort(sortBy(pods))
+		return pods[0], len(podList.Items), nil
+	}
+
+	// Watch for ready pods.
+	options.ResourceVersion = podList.ResourceVersion
+	w, err := client.Pods(namespace).Watch(options)
+	if err != nil {
+		return nil, len(podList.Items), err
+	}
+	defer w.Stop()
+
+	condition := func(event watch.Event) (bool, error) {
+		pod, ok := event.Object.(*api.Pod)
+		if !ok {
+			return false, fmt.Errorf("%#v is not a pod event", event)
 		}
+		return api.IsPodReady(pod), nil
```

Comment:
You are right, I didn't consider this once I factored out the readiness check above. Making sure that the pod has been scheduled would be enough I guess?